### PR TITLE
feat: all custom envs are compiled during installation

### PR DESCRIPTION
**Pull request by zkochan**
## Proposed Changes

-
-
-

### CODE REVIEW

File: scopes/workspace/install/install.main.runtime.ts [hunk 1]
```diff
@@ -174,22 +163,79 @@ export class InstallMain {
...
     };
-    await installer.install(
-      this.workspace.path,
-      mergedRootPolicy,
-      compDirMap,
-      { installTeambitBit: false },
-      pmInstallOptions
-    );
+    // TODO: pass get install options
+    const installer = this.dependencyResolver.getInstaller({});
```
Comment (GiladShoham) [inline scopes/workspace/install/install.main.runtime.ts:167]: @zkochan do we need to implement this todo before the merge?

File: scopes/workspace/install/install.main.runtime.ts [hunk 2]
```diff
@@ -404,6 +462,18 @@ export class InstallMain {
...
 
+interface AllWorkspaceManifests {
+  componentsManifests: Record<string, ProjectManifest>;
+  workspaceManifest: ProjectManifest;
+}
+
+function isManifestsEqual(prev: AllWorkspaceManifests, next: AllWorkspaceManifests): boolean {
+  return (
+    isEqual(prev.componentsManifests, next.componentsManifests) &&
+    JSON.stringify(prev.workspaceManifest) === JSON.stringify(next.workspaceManifest)
```
Comment (GiladShoham) [inline scopes/workspace/install/install.main.runtime.ts:473]: why different types of compare?
also, I wonder if the stringify compare might be affected here by the order of the keys which might be different maybe between iterations?

File: scopes/workspace/install/install.main.runtime.ts [hunk 3]
```diff
@@ -174,22 +163,79 @@ export class InstallMain {
...
+      linkTeambitBit: false,
+      legacyLink: true,
+      linkCoreAspects: false,
+      linkDepsResolvedFromEnv: !hasRootComponents,
       linkNestedDepsInNM: !this.workspace.isLegacy && !hasRootComponents,
+    };
+    /* eslint-disable no-await-in-loop */
+    do {
+      this.workspace.consumer.componentLoader.clearComponentsCache();
+      this.workspace.clearCache();
```
Comment (GiladShoham) [inline scopes/workspace/install/install.main.runtime.ts:190]: I wonder if we can improve the cache cleaning here somehow.
as this is a very expensive op (not the clean, the reload again once it's cleaned)
maybe we should somehow check what should be cleaned by checking if the component's env wasn't loaded in the previous iteration or something like this.
we can keep this optimization to a different PR, but may be worth discussing it also with @davidfirst

File: scopes/workspace/install/install.main.runtime.ts (hunk 1 above)
Comment (zkochan) [reply scopes/workspace/install/install.main.runtime.ts:167]: This TODO was there before my changes. I just moved the code.

File: scopes/workspace/install/install.main.runtime.ts (hunk 3 above)
Comment (zkochan) [reply scopes/workspace/install/install.main.runtime.ts:190]: actually, it seems to work without clearing the cache. So I'll remove these lines

File: scopes/workspace/install/install.main.runtime.ts (hunk 2 above)
Comment (zkochan) [reply scopes/workspace/install/install.main.runtime.ts]: Changed to isEqual in both cases. isEqual doesn't care about the order of keys.

**Review by GiladShoham — APPROVED** [review]
